datafusion/optimizer/src/analyzer/mod.rs

```diff
@@ -27,8 +27,16 @@ use log::{debug, trace};
 use std::sync::Arc;
 use std::time::Instant;
 
-/// `AnalyzerRule` transforms the unresolved ['LogicalPlan']s and unresolved ['Expr']s into
-/// the resolved form.
+/// [`AnalyzerRule`]s transform [`LogicalPlan`]s in some way to make
+/// the plan valid prior to the rest of the DataFusion optimization process.
+///
+/// For example, it may resolve [`Expr]s into more specific forms such
+/// as a subquery reference, to do type coercion to ensure the types
+/// of operands are correct.
+///
+/// This is different than an [`OptimizerRule`](crate::OptimizerRule)
+/// which should preserve the semantics of the LogicalPlan but compute
+/// it the same result in some more optimal way.
 pub trait AnalyzerRule {
     /// Rewrite `plan`
     fn analyze(&self, plan: &LogicalPlan, config: &ConfigOptions) -> Result<LogicalPlan>;
```

Comment:
While reading this I wonder why it takes `plan: &LogicalPlan` rather than `plan: LogicalPlan` -- given that `LogicalPlan` is returned this interface requires `clone`ing the LogicalPlan on each step

Reply:
My bad, I just copy it from Logical optimizer rule. I think for Analyzer, we should change it to `plan: LogicalPlan`.

Reply:
Proposed PR: https://github.com/apache/arrow-datafusion/pull/5728